src/server/transaction.cc

```diff
@@ -1135,88 +1050,39 @@ OpArgs Transaction::GetOpArgs(EngineShard* shard) const {
-  DCHECK(!IsAtomicMulti());
-  DCHECK_EQ(txid_, 0u);
-  DCHECK(shard_data_.size() == 1u || multi_->mode == NON_ATOMIC);
-  DCHECK_NE(unique_shard_id_, kInvalidSid);
-
-  auto mode = LockMode();
-  auto lock_args = GetLockArgs(shard->shard_id());
-
-  auto& sd = shard_data_[SidToId(unique_shard_id_)];
-  DCHECK_EQ(TxQueue::kEnd, sd.pq_pos);
-
-  bool shard_unlocked = shard->shard_lock()->Check(mode);
-  bool keys_unlocked = shard->db_slice().Acquire(mode, lock_args);
-  bool quick_run = shard_unlocked && keys_unlocked;
-  bool continue_scheduling = !quick_run;
-
-  sd.local_mask |= KEYLOCK_ACQUIRED;
-
-  // Fast path. If none of the keys are locked, we can run briefly atomically on the thread
-  // without acquiring them at all.
-  if (quick_run) {
-    CHECK(sd.is_armed.exchange(false, memory_order_relaxed));
-    auto result = RunQuickie(shard);
-    local_result_ = result.status;
-
-    if (result.flags & RunnableResult::AVOID_CONCLUDING) {
-      // If we want to run again, we have to actually schedule this transaction
-      DCHECK(!sd.is_armed.load(memory_order_relaxed));
-      continue_scheduling = true;
-    } else {
-      LogAutoJournalOnShard(shard, result);
-      shard->db_slice().Release(mode, lock_args);
-      sd.local_mask &= ~KEYLOCK_ACQUIRED;
-    }
-  }
-
-  // Slow path. Some of the keys are locked, so we schedule on the transaction queue.
-  if (continue_scheduling) {
-    coordinator_state_ |= COORD_SCHED;                  // safe because single shard
-    txid_ = op_seq.fetch_add(1, memory_order_relaxed);  // -
-    sd.pq_pos = shard->txq()->Insert(this);
-
-    DVLOG(1) << "Rescheduling " << DebugId() << " into TxQueue of size " << shard->txq()->size();
-
-    // If there are blocked transactons waiting for these tx keys, we add this transaction
-    // to the tx-queue (these keys will be contended). This happen even if the queue is empty.
-    // In that case we must poll the queue, because there will be no other callback trigerring the
-    // queue before us.
-    shard->PollExecution("schedule_unique", nullptr);
-  }
-
-  return quick_run;
-}
-
 // This function should not block since it's run via RunBriefInParallel.
-bool Transaction::ScheduleInShard(EngineShard* shard) {
+bool Transaction::ScheduleInShard(EngineShard* shard, bool can_run_immediately) {
   ShardId sid = SidToId(shard->shard_id());
   auto& sd = shard_data_[sid];
 
   DCHECK(sd.local_mask & ACTIVE);
   DCHECK_EQ(sd.local_mask & KEYLOCK_ACQUIRED, 0);
-  sd.local_mask &= ~OUT_OF_ORDER;
-
-  // If a more recent transaction already commited, we abort
-  if (shard->committed_txid() >= txid_)
-    return false;
+  sd.local_mask &= ~(OUT_OF_ORDER | RAN_IMMEDIATELY);
 
   TxQueue* txq = shard->txq();
   KeyLockArgs lock_args;
   IntentLock::Mode mode = LockMode();
   bool lock_granted = false;
 
+  // If a more recent transaction already commited, we abort
+  if (txid_ > 0 && shard->committed_txid() >= txid_)
```

Comment:
should we initialize tx_ to `UINT64_MAX`?

Reply:
Better not, we have lots of logic relying on txid=0